src/main/java/agh/ics/oop/gui/App.java

```diff
@@ -77,17 +75,65 @@ private void updateColRowDimensions() {
         }
         if (numOfRows != this.numOfRows) {
             this.numOfRows = numOfRows;
-            int rowHeight = HEIGHT/numOfRows;
+            int rowHeight = (HEIGHT-2*FONTSIZE)/numOfRows;
             this.gridPane.getRowConstraints().clear();
             IntStream.range(0, numOfRows).forEach(i -> this.gridPane.getRowConstraints().add(new RowConstraints(rowHeight)));
         }
+        this.gridPane.setGridLinesVisible(true);
+    }
+
+    private void renderMap() {
+        this.gridPane.getChildren().clear();
+        this.initGridPane();
+        this.updateColRowDimensions();
+        map.mapElements().forEach(element -> {
+            // check if resource exists
+            GuiElementBox guiElement = this.cachedResources.get(element.hashCode());
```

Comment:
tutaj praktycznie nigdy nie dostaniemy żadnego węzła, bo hashe elementów zmieniają się w momencie gdy np. zmieniamy im pozycje/obracamy je. To w sumie fajny przykład, dlaczego kluczami mapy powinny być obiekty immutable (i ogólnie, czemu modele immutable są bezpieczniejsze). 

Btw, nawet jakby ten cache działał to 1) lepiej jako kluczy używać pełnych obiektów niż robić robotę za hashmape i trzymać tam hashe 2) w tym przypadku niespecjalnie ten cache nam pomaga, patrz uwaga do GUI elementu.